### PR TITLE
Fix ExistentialSpecializer bugs for swiftpm and enable existential-specializer for sil-opt

### PR DESCRIPTION
**Pull request by rajbarik**

### CODE REVIEW

**Comment by rajbarik** [comment]
@slavapestov test cases added and made into a single commit. Can you check this one?

**Comment by rajbarik** [comment]
@slavapestov can you check this one now?

File: lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp [hunk 1]
```diff
@@ -101,7 +101,8 @@ static bool findConcreteType(ApplySite AI, int ArgIdx, CanType &ConcreteType) {
     SILValue InitExistential =
         findInitExistentialFromGlobalAddrAndApply(GAI, AI, ArgIdx);
     /// If the Arg is already init_existential, return the concrete type.
-    if (findConcreteTypeFromInitExistential(InitExistential, ConcreteType)) {
+    if (InitExistential &&
```
Comment (rajbarik) [inline lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp]: This fixes the  case where global_addr analysis can not find an init_existential. See "global_addr_init" test case below.

File: lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp [hunk 2]
```diff
@@ -257,6 +258,12 @@ bool ExistentialSpecializer::canSpecializeCalleeFunction(ApplySite &Apply) {
   if (Callee->getInlineStrategy() == Inline_t::AlwaysInline)
     return false;
 
+  /// Ignore externally linked functions with public_external or higher
+  /// linkage.
+  if (isAvailableExternally(Callee->getLinkage())) {
```
Comment (rajbarik) [inline lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp]: This fixes the swiftpm bug related to external linkage functions causing name mangling issues.

File: lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp [hunk 3]
```diff
@@ -192,13 +189,11 @@ void ExistentialSpecializerCloner::cloneAndPopulateFunction() {
...
   llvm::SmallPtrSet<ReturnInst *, 4> ReturnInsts;
-  if (MissingDestroyUse) {
-    /// Find the set of return instructions in a function.
-    for (auto &BB : NewF) {
-      TermInst *TI = BB.getTerminator();
-      if (auto *RI = dyn_cast<ReturnInst>(TI)) {
-        ReturnInsts.insert(RI);
-      }
+  /// Find the set of return instructions in a function.
+  for (auto &BB : NewF) {
```
Comment (rajbarik) [inline lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp]: This fixes the old code that was inserting dealloc_stack after destroy_addr which is wrong. dealloc_stack are now inserted just before return instructions. See test case: $s7dealloc12wrap_foo_ncp1a1bSiAA1P_pz_AaE_pztFTf4ee_n below.

File: include/swift/AST/SILOptions.h [hunk 4]
```diff
@@ -51,7 +51,7 @@ class SILOptions {
   bool RemoveRuntimeAsserts = false;
 
   /// Enable existential specializer optimization.
-  bool ExistentialSpecializer = false;
+  bool ExistentialSpecializer = true;
```
Comment (rajbarik) [inline include/swift/AST/SILOptions.h]: default enable of ExistentialSpecializer.

**Comment by slavapestov** [comment]
Can you file a separate PR with the bug fixes before we enable this by default?

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by rajbarik** [comment]
I notice that you ran tests. Do you still suggest I need to separate them into two PRs -- one with just ExistentialSpecializer set to true in SILOptions (this exists here  https://github.com/apple/swift/pull/19820) and another with all bug fixes (I can modify this one to remove the set to ExistentialSpecializer=true)? 

A question: if I do not set "ExistentialSpecializer=true" in SILOptions.h, ExistentialSpecializer does not get invoked with sil-opt using the "-existential-specializer" (aka, "// RUN: %target-sil-opt -assume-parsing-unqualified-ownership-sil -enable-sil-verify-all %s -existential-specializer   | %FileCheck %s") used in my test case existential_transform_extras.sil. How do I enable this at sil-opt level?

**Comment by rajbarik** [comment]
@slavapestov anything for me here?

**Comment by slavapestov** [comment]
Since we have performance regressions here, I’d like you to merge the bug fixes separately. Did you figure out how to get the tests to pass even when the flag is off?

**Comment by rajbarik** [comment]
Looking at Performance regression with -O for StringEqualPointerComparison:
This test has nothing to do with Protocols.

**Comment by rajbarik** [comment]
Similarly, none of the regressed one have anything to do with protocols. 

What am I missing?

File: tools/sil-opt/SILOpt.cpp [hunk 5]
```diff
@@ -167,6 +167,11 @@ static llvm::cl::opt<int>
 SILInlineThreshold("sil-inline-threshold", llvm::cl::Hidden,
                    llvm::cl::init(-1));
 
+static llvm::cl::opt<bool>
+SILExistentialSpecializer("enable-sil-existential-specializer",
```
Comment (rajbarik) [inline tools/sil-opt/SILOpt.cpp]: Add a flag to enable and disable existential-specializer from sil-opt.

File: tools/sil-opt/SILOpt.cpp [hunk 6]
```diff
@@ -328,6 +333,7 @@ int main(int argc, char **argv) {
   // Setup the SIL Options.
   SILOptions &SILOpts = Invocation.getSILOptions();
   SILOpts.InlineThreshold = SILInlineThreshold;
+  SILOpts.ExistentialSpecializer = SILExistentialSpecializer;
```
Comment (rajbarik) [inline tools/sil-opt/SILOpt.cpp]: Store it in the module's Options.

**Comment by rajbarik** [comment]
OK. I committed some more changes to make it independent:

1) It does not default enable ExistentialSpecializer option from the compiler. We should use the old PR https://github.com/apple/swift/pull/19820 for turning on by default.
2) Introduced a enable flag in sil-opt to be able to test it using sil-opt.

**Comment by rajbarik** [comment]
@slavapestov can we run tests on this one and accept these changes quickly? 

The default enabler of ExistentialSpecializer is a separate PR at https://github.com/apple/swift/pull/19820

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Pull request merged**